grading/execute.cpp

```diff
@@ -39,150 +39,51 @@ extern const int CPU_TO_WALLCLOCK_TIME_BUFFER;  // defined in default_config.h
-
-    // for Operating Systems
-    { "gcc",                     "/usr/bin/gcc" },
-    { "strings",                 "/usr/bin/strings" },
-
-    // for Programming Languages
-    { "swipl",                   "/usr/bin/swipl" },
-    { "plt-r5rs",                "/usr/bin/plt-r5rs" },
-    { "ozc",                     "/usr/bin/ozc" },
-    { "ozengine",                "/usr/bin/ozengine" },
-
-    // for Program Analysis course
-    { "ghc",                     "/usr/bin/ghc" },
-    { "ocaml",                   "/usr/bin/ocaml" },
-    { "ocamllex",                "/usr/bin/ocamllex" },
-    { "ocamlyacc",               "/usr/bin/ocamlyacc" },
-    { "z3",                      SUBMITTY_INSTALL_DIRECTORY+"/tools/z3" },
-
-    // for Cmake & Make
-    { "cmake",                   "/usr/bin/cmake" },
-    { "make",                    "/usr/bin/make" },
-
-    // for Network Programming
-    { "timeout",                 "/usr/bin/timeout" },
-    { "mpicc.openmpi",           "/usr/bin/mpicc.openmpi" },
-    { "mpirun.openmpi",          "/usr/bin/mpirun.openmpi" },
-    { "mpirun",                  "/usr/local/mpich-3.2/bin/mpirun"},
-    { "mpicc",                   "/usr/local/mpich-3.2/bin/mpicc"},
-    { "expect",                  "/usr/bin/expect" },
-    { "sleep",                   "/bin/sleep" },
-
-    // for Distributed Systems
-    { "script",                  "/usr/bin/script" },
-
-    // for LLVM / Compiler class
-    { "lex",                     "/usr/bin/lex" },
-    { "flex",                    "/usr/bin/flex" },
-    { "yacc",                    "/usr/bin/yacc" },
-    { "bison",                   "/usr/bin/bison" },
-
-    // for graphics/window interaction
-    { "scrot",                   "/usr/bin/scrot"}, //screenshot utility
-    { "xdotool",                 "/usr/bin/xdotool"}, //keyboard/mouse input
-    { "wmctrl",                  "/usr/bin/wmctrl"}, //bring window into focus
-    { "xwininfo",                "/usr/bin/xwininfo"}, // get info about window
-
-    // for Debugging
-    { "strace",                  "/usr/bin/strace" },
-
-    //Matlab
-    { "matlab",                  "/usr/local/bin/matlab" }
-
-  };
-
-  if(running_in_docker){
-    allowed_system_programs.insert({"bash", "/bin/bash"});
-    allowed_system_programs.insert({"php",  "/usr/bin/php"});
-  }
-  // find full path name
-  std::map<std::string,std::string>::const_iterator itr = allowed_system_programs.find(program);
-  if (itr != allowed_system_programs.end()) {
-    full_path_executable = itr->second;
-    return true;
-  }
-
-  // did they already use the full path name?
-  for (itr = allowed_system_programs.begin(); itr != allowed_system_programs.end(); itr++) {
-    if (itr->second == program) {
-      full_path_executable = program;
-      return true;
+bool system_program(const std::string &program, std::string &full_path_executable, const bool running_in_docker)
```

Comment:
Will `system_program` be called more than once during the execution of this program? If yes, then it might be good to define `allowed_system_programs` early on and use the same on every call instead of creating a new json object on every call.

Reply:
I also had the same idea, but I thought since this is a secure piece of information we wouldn't want a object passing around functions. Not sure about my argument though.

Reply:
That shouldn't be a problem, but this argument is convincing enough to err on the side of caution.